src/cloudai/systems/kubernetes/kubernetes_system.py

```diff
@@ -0,0 +1,609 @@
+                    f"Error code: {e.status}. Message: {e.reason}. Please check the job name, namespace, and "
+                    "Kubernetes API server."
+                )
+                logging.error(error_message)
+                raise
+
+    def _is_batch_job_running(self, job_namespace: str, job_name: str) -> bool:
+        """
+        Check if a batch job is currently running.
+
+        Args:
+            job_namespace (str): The namespace of the batch job.
+            job_name (str): The name of the batch job.
+
+        Returns:
+            bool: True if the batch job is running, False if the job has completed or is not found.
+        """
+        try:
+            k8s_job: Any = self.batch_v1.read_namespaced_job_status(name=job_name, namespace=job_namespace)
+
+            if not (hasattr(k8s_job, "status") and hasattr(k8s_job.status, "conditions")):
+                logging.debug(
+                    f"Job '{job_name}' in namespace '{job_namespace}' does not have expected status attributes."
+                )
+                return False
+
+            conditions = k8s_job.status.conditions or []
+
+            for condition in conditions:
+                if condition.type == "Complete" and condition.status == "True":
+                    return False
+                if condition.type == "Failed" and condition.status == "True":
+                    return False
+
+            return any(condition.type == "Created" and condition.status == "True" for condition in conditions)
+
+        except ApiException as e:
+            if e.status == 404:
+                logging.debug(
+                    f"Batch job '{job_name}' not found in namespace '{job_namespace}'."
+                    "It may have completed and been removed from the system."
+                )
+                return False
+            else:
+                logging.error(
+                    f"Error occurred while retrieving status for batch job '{job_name}' in namespace "
+                    "'{job_namespace}'."
+                    f"Error code: {e.status}. Message: {e.reason}. Please check the job name, namespace, "
+                    "and Kubernetes API server."
+                )
+                raise
+
+    def kill(self, job: BaseJob) -> None:
+        """
+        Terminate a Kubernetes job.
+
+        Args:
+            job (BaseJob): The job to be terminated.
+        """
+        k_job: KubernetesJob = cast(KubernetesJob, job)
+        self.delete_job(k_job.get_namespace(), k_job.get_name(), k_job.get_kind())
+
+    def delete_job(self, namespace: str, job_name: str, job_kind: str) -> None:
+        """
+        Delete a job in the specified namespace.
+
+        Args:
+            namespace (str): The namespace of the job.
+            job_name (str): The name of the job.
+            job_kind (str): The kind of the job ('MPIJob' or 'Job').
+        """
+        if "mpijob" in job_kind.lower():
```

Comment:
We basically have two classes here: K8s MPI and K8s Batch. But it is only related to Job runnings part, the system itself it the same AFAIU. Moving job handling part into two separate classes would benefit readability and maintainability.

I guess this is not the task for this PR, but I don't see tests for the current implementation. Testing this code will be much more difficult compared to designed where system state and job handling responsibilities are split. Let's discuss it offline.

Reply:
> We basically have two classes here: K8s MPI and K8s Batch. But it is only related to Job runnings part, the system itself it the same AFAIU. Moving job handling part into two separate classes would benefit readability and maintainability.

I agree, I will update it accordingly.

> I guess this is not the task for this PR, but I don't see tests for the current implementation. Testing this code will be much more difficult compared to designed where system state and job handling responsibilities are split. Let's discuss it offline.

Testing the Kubernetes system class with unit tests is not efficient, and I think testing requires a real system. Do you have any suggestions for testing the class using unit tests? I don't believe the return on investment is high enough if we should test this class with unit tests.

Reply:
> Testing the Kubernetes system class with unit tests is not efficient, and I think testing requires a real system. Do you have any suggestions for testing the class using unit tests?

I don't believe in "real platform testing". This essentially mean "our users will test it in prod". No way we should do that.

K8s is a system and we interact with it in a limited way. When we do that, there are limited outcomes, say we call a function. It can return a value or raise an exception. We know that and we test both using mocked K8s. Next, returned value can be valid or not. We test that too.

When we create a PR, we should be sure it is not breaking anything. It is true for the `main` branch state right now. This PR changes it by introducing the whole System without any testing. I disagree with this approach.

Reply:
It makes sense. My question is, how can we test without having a real system? I don't have much experience or a clear idea about this.